### PR TITLE
Persist muted diagnostics & show muted in settings

### LINKED ISSUE #2968: Muted diagnostics don't stay muted
When I move away from the Diagnostics tab, any muted diagnostics revert to unmuted state, and are visible again when I move back to the tab.

The mute state should persist permanently, so even stopping and restarting the app should not unmute any diagnostics.

### PR DESCRIPTION
**Pull request by rak-phillip**
This persists choices made for muting individual diagnostics and showing muted diagnostics in the settings object. In order to achieve this, we added a new `diagnostics` property to the settings object as well as updating the `diagnostics` and `preferences` stores so that they save these choices while muting.

closes #2968

### CODE REVIEW

File: src/components/DiagnosticsBody.vue [hunk 1]
```diff
@@ -83,7 +84,7 @@ export default Vue.extend({
       this.$store.dispatch('diagnostics/updateDiagnostic', { isMuted, row });
     },
     toggleMute() {
-      this.showMuted = !this.showMuted;
+      this.$store.dispatch('preferences/toggleShowMuted');
```
Comment (mook-as) [inline src/components/DiagnosticsBody.vue:87]: Hmm, from a user point of view, shouldn't this just be set to whatever the current value of the toggle switch is, instead of toggling? (In case the pref was changed in the background and the front end wasn't updated somehow.)

File: src/main/commandServer/settingsValidator.ts [hunk 2]
```diff
@@ -32,6 +32,8 @@ type SettingsValidationMapEntry<T> = {
   ValidatorFunc<T[k], T[k]> :
   T[k] extends Record<string, infer V> ?
   SettingsValidationMapEntry<T[k]> | ValidatorFunc<T[k], Record<string, V>> :
+  T[k] extends Array<infer V> ?
+  ValidatorFunc<T[k], Array<V>> :
```
Comment (mook-as) [inline src/main/commandServer/settingsValidator.ts:36]: This ended up being unnecessary (because we ended up using a mapping instead of an array), but I guess it doesn't hurt to keep it?

File: src/main/commandServer/settingsValidator.ts [hunk 3]
```diff
@@ -229,6 +235,8 @@ export default class SettingsValidator {
     return currentValue !== desiredEngine;
   }
 
+  protected checkMutedDiagnostics = this.checkWSLIntegrations;
```
Comment (mook-as) [inline src/main/commandServer/settingsValidator.ts:238]: We may need to rename the method to make it clearer that it's shared… (so we don't end up changing the implementation and forget that it also applies to mute settings).

File: src/store/diagnostics.ts [hunk 4]
```diff
@@ -14,6 +14,23 @@ interface DiagnosticsState {
...
+/**
+ * Updates the muted property for diagnostic results.
+ * @param checks A collection of diagnostic results that require muting.
+ * @param mutedChecks A collection of key, value pairs that contains a key of
+ * the ID for the diagnostic and a boolean value for muting the result.
+ * @returns A collection of diagnostic results with an updated muted property.
+ */
+const mapMutedDiagnostics = (checks: DiagnosticsResult[], mutedChecks: Record<string, boolean>) => {
+  return checks.map((check) => {
+    if (Object.keys(mutedChecks).includes(check.id)) {
```
Comment (mook-as) [inline src/store/diagnostics.ts:26]:
```suggestion
    if (check.id in mutedChecks) {
```

File: src/store/diagnostics.ts [hunk 5]
```diff
@@ -14,6 +14,23 @@ interface DiagnosticsState {
...
+      check.mute = mutedChecks[check.id];
```
Comment (mook-as) [inline src/store/diagnostics.ts:27]: We're modifying this in-place; is that okay?

File: src/store/diagnostics.ts [hunk 6]
```diff
@@ -97,12 +122,29 @@ export const actions = {
...
     rowToUpdate.mute = isMuted;
 
+    await dispatch(
+      'preferences/updatePreferencesData',
+      {
+        property: 'diagnostics.mutedChecks',
+        value:    {
+          ...rootState.preferences.preferences.diagnostics.mutedChecks,
+          [rowToUpdate.id]: isMuted,
+        } as Record<string, boolean>,
```
Comment (mook-as) [inline src/store/diagnostics.ts:132]: Any reason we can't just do
```
property: `diagnostics.mutedChecks.${rowToUpdate.id}`
```

File: src/store/diagnostics.ts (hunk 5 above)
Comment (rak-phillip) [reply src/store/diagnostics.ts:27]: good catch, this is not okay..

File: src/main/commandServer/settingsValidator.ts (hunk 2 above)
Comment (rak-phillip) [reply src/main/commandServer/settingsValidator.ts:36]: I guess this isn't the PR to add this in and as it's not used so I guess it's technically cruft.. removed. We can easily add this back if needed.

File: src/store/diagnostics.ts (hunk 6 above)
Comment (rak-phillip) [reply src/store/diagnostics.ts]: I mean, you can if want to express the same thing in a more succinct way 😆

File: src/main/commandServer/settingsValidator.ts (hunk 3 above)
Comment (rak-phillip) [reply src/main/commandServer/settingsValidator.ts]: We discussed offline and decided to rename `checkWSLIntegrations` to `checkBooleanMapping`

File: src/store/diagnostics.ts [hunk 7]
```diff
@@ -14,6 +14,19 @@ interface DiagnosticsState {
...
+ * the ID for the diagnostic and a boolean value for muting the result.
+ * @returns A collection of diagnostic results with an updated muted property.
+ */
+const mapMutedDiagnostics = (checks: DiagnosticsResult[], mutedChecks: Record<string, boolean>) => {
+  return checks.map((check) => {
+    return check.id in mutedChecks ? { ...check, mute: mutedChecks[check.id] } : check;
+  });
+};
+
 export const state: () => DiagnosticsState = () => (
```
Comment (mook-as) [inline src/store/diagnostics.ts:30]: DiagnosticsState should now be something like:
```ts

interface DiagnosticsState {
  diagnostics: Array<DiagnosticsResult & { mute: boolean }>,
  timeLastRun: Date;
  inError: boolean;
}
```

File: src/store/diagnostics.ts [hunk 8]
```diff
@@ -14,6 +14,19 @@ interface DiagnosticsState {
...
+ * Updates the muted property for diagnostic results.
+ * @param checks A collection of diagnostic results that require muting.
+ * @param mutedChecks A collection of key, value pairs that contains a key of
+ * the ID for the diagnostic and a boolean value for muting the result.
+ * @returns A collection of diagnostic results with an updated muted property.
+ */
+const mapMutedDiagnostics = (checks: DiagnosticsResult[], mutedChecks: Record<string, boolean>) => {
+  return checks.map((check) => {
+    return check.id in mutedChecks ? { ...check, mute: mutedChecks[check.id] } : check;
+  });
```
Comment (mook-as) [inline src/store/diagnostics.ts:27]: There's no reason not to always have `mute` on the result:
```suggestion
  return checks.map((check) => ({ ...check, mute: !!mutedChecks[check.id] }));
```

File: src/store/diagnostics.ts [hunk 9]
```diff
@@ -97,12 +118,26 @@ export const actions = {
...
+      { root: true },
+    );
+
     commit('SET_DIAGNOSTICS', diagnostics);
   },
 };
 
 export const getters: GetterTree<DiagnosticsState, DiagnosticsState> = {
-  diagnostics(state: DiagnosticsState) {
+  diagnostics(state: DiagnosticsState, _getters) {
```
Comment (mook-as) [inline src/store/diagnostics.ts:140]: Huh, did we need `_getters` for some reason? (It's probably fine to leave the change in, just curious.)

File: src/main/commandServer/settingsValidator.ts [hunk 10]
```diff
@@ -259,7 +263,7 @@ export default class SettingsValidator {
     return false;
   }
 
-  protected checkWSLIntegrations(currentValue: Record<string, boolean>, desiredValue: Record<string, boolean>, errors: string[], fqname: string): boolean {
```
Comment (mook-as) [inline src/main/commandServer/settingsValidator.ts:262]: This probably needs a comment because it's not clear from the function name what it does. (I realize that that is my fault, since I helped with picking the name…)

Even more my fault: all the other functions are `checkX`, instead of `validateX`…

File: src/store/preferences.ts [hunk 11]
```diff
@@ -206,6 +206,22 @@ export const actions = {
 
     return severities;
   },
+  async toggleShowMuted({ dispatch, rootState }: PrefActionContext, isMuted: boolean) {
```
Comment (mook-as) [inline src/store/preferences.ts:209]: This is a `set` rather than a `toggle` (because it doesn't depend on the current state), right?

File: src/store/preferences.ts [hunk 12]
```diff
@@ -206,6 +206,22 @@ export const actions = {
...
+      'preferences/updatePreferencesData',
+      {
+        property: 'diagnostics.showMuted',
+        value:    isMuted,
+      },
+      { root: true },
+    );
+
+    await dispatch(
+      'preferences/commitPreferences',
```
Comment (mook-as) [inline src/store/preferences.ts:220]: Will this accidentally commit other, unrelated changes? For example, if I also have the prefs window open and managed to change some things…
(I guess that's impossible due to the prefs window being modal?)

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/store/diagnostics.ts (hunk 9 above)
Comment (rak-phillip) [reply src/store/diagnostics.ts:140]: No, I had a different solution for the getters in a previous commit, but I missed the unused param after cleaning up because of the `_`

File: src/store/diagnostics.ts (hunk 7 above)
Comment (mook-as) [reply src/store/diagnostics.ts:30]: Phil brought up that this makes no sense — `DiagnosticsResult` _already_ has a `mute` property.

File: src/store/preferences.ts (hunk 12 above)
Comment (rak-phillip) [reply src/store/preferences.ts]: This is a very good catch. I was originally trying to avoid changing `commitPreferences` but I think it makes sense that we do so in order to avoid this. 

I've added an optional `payload` parameter that we can use to update only the props that interest us instead of the entirety of the preferences state.

File: src/main/commandServer/settingsValidator.ts (hunk 10 above)
Comment (rak-phillip) [reply src/main/commandServer/settingsValidator.ts:262]: Agreed, best to be consistent here. I added some comments, let me know what you think.

File: src/components/DiagnosticsBody.vue [hunk 13]
```diff
@@ -83,7 +84,7 @@ export default Vue.extend({
       this.$store.dispatch('diagnostics/updateDiagnostic', { isMuted, row });
     },
     toggleMute() {
-      this.showMuted = !this.showMuted;
+      this.$store.dispatch('preferences/setShowMuted', !this.showMuted as boolean);
```
Comment (mook-as) [inline src/components/DiagnosticsBody.vue:87]: I'm confused — shouldn't `!x` always be a boolean?
(If TypeScript complains, we can't do much about it I guess…)

File: src/main/commandServer/settingsValidator.ts [hunk 14]
```diff
@@ -259,7 +263,13 @@ export default class SettingsValidator {
...
+  /**
+   * Ensures that settings that are object that adhere to their type of
```
Comment (ericpromislow) [inline src/main/commandServer/settingsValidator.ts:267]: Not really a sentence

File: src/store/diagnostics.ts [hunk 15]
```diff
@@ -3,6 +3,7 @@ import { GetterTree } from 'vuex';
 
 import { ActionContext, MutationsType } from './ts-helpers';
 
+import { Settings } from '@/config/settings';
```
Comment (mook-as) [inline src/store/diagnostics.ts:6]:
```suggestion
import type { Settings } from '@/config/settings';
```
Don't want to end up pulling in weird things into the renderer process…

File: src/store/diagnostics.ts [hunk 16]
```diff
@@ -97,12 +117,23 @@ export const actions = {
 
     rowToUpdate.mute = isMuted;
 
+    const mutedChecks = { ...rootState.preferences.preferences.diagnostics.mutedChecks, [rowToUpdate.id]: isMuted };
+
+    await dispatch(
+      'preferences/commitPreferences',
+      {
+        ...rootState.credentials.credentials as ServerState,
+        payload: { diagnostics: { mutedChecks } } as Partial<Settings>,
```
Comment (mook-as) [inline src/store/diagnostics.ts:126]: Huh, wouldn't that need to be `RecursivePartial` instead of `Partial`? (`Partial` means it expects all children to be fully there.)

File: src/store/preferences.ts [hunk 17]
```diff
@@ -25,6 +25,10 @@ interface PreferencesState {
   preferencesError: string;
 }
 
+interface CommitArgs extends ServerState {
+  payload?: Partial<Settings>;
```
Comment (mook-as) [inline src/store/preferences.ts:29]:
```suggestion
  payload?: RecursivePartial<Settings>;
```
`Partial<{a: {b: 1}}>` does not accept `{a: {}}`, because `Partial` is not recursive (it expects any children that exist to be the same type, and not `Partial` itself).

File: src/store/preferences.ts [hunk 18]
```diff
@@ -100,8 +104,12 @@ export const actions = {
...
     dispatch('preferences/initializePreferences', settings, { root: true });
   },
-  async commitPreferences({ state, dispatch }: PrefActionContext, args: ServerState) {
-    const { port, user, password } = args;
+  async commitPreferences({ state, dispatch }: PrefActionContext, args: CommitArgs) {
+    const {
+      port, user, password, payload,
+    } = args;
+
+    console.debug({ payload });
```
Comment (mook-as) [inline src/store/preferences.ts:112]: (This needs to be removed before merge)

File: src/store/preferences.ts [hunk 19]
```diff
@@ -111,7 +119,7 @@ export const actions = {
           Authorization:  `Basic ${ window.btoa(`${ user }:${ password }`) }`,
           'Content-Type': 'application/x-www-form-urlencoded',
         }),
-        body: JSON.stringify(state.preferences),
+        body: JSON.stringify(payload || state.preferences),
```
Comment (mook-as) [inline src/store/preferences.ts:122]:
```suggestion
        body: JSON.stringify(payload ?? state.preferences),
```
Doesn't make a difference here, but seems clearer.

File: src/store/preferences.ts [hunk 20]
```diff
@@ -206,6 +214,16 @@ export const actions = {
 
     return severities;
   },
+  async setShowMuted({ dispatch, rootState }: PrefActionContext, isMuted: boolean) {
+    await dispatch(
+      'preferences/commitPreferences',
+      {
+        ...rootState.credentials.credentials as ServerState,
+        payload: { diagnostics: { showMuted: isMuted } } as Partial<Settings>,
```
Comment (mook-as) [inline src/store/preferences.ts:222]:
```suggestion
        payload: { diagnostics: { showMuted: isMuted } } as RecursivePartial<Settings>,
```
(Actually, with the fix for the definition of `CommitArgs` the cast should be unnecessary.)

**Review by mook-as — CHANGES_REQUESTED** [review]
<3

File: src/components/DiagnosticsBody.vue (hunk 13 above)
Comment (rak-phillip) [reply src/components/DiagnosticsBody.vue:87]: Yes, we get typescript complaints because `mapGetters` doesn't really preserve type information..

File: src/store/preferences.ts (hunk 20 above)
Comment (rak-phillip) [reply src/store/preferences.ts:222]: This is true, removed

Comment (rak-phillip) [reply src/store/preferences.ts]: NVM, maybe not. Looks like my editor wasn't picking up on the deprecation warning, but the linter definitely is..

File: src/store/diagnostics.ts [hunk 21]
```diff
@@ -97,12 +118,23 @@ export const actions = {
 
     rowToUpdate.mute = isMuted;
 
+    const mutedChecks = { ...rootState.preferences.preferences.diagnostics.mutedChecks, [rowToUpdate.id]: isMuted };
+
+    await dispatch(
+      'preferences/commitPreferences',
+      {
+        ...rootState.credentials.credentials as ServerState,
+        payload: { diagnostics: { mutedChecks } } as RecursivePartial<Settings>,
```
Comment (mook-as) [inline src/store/diagnostics.ts:127]: Just realized that `{ diagnostics: { mutedChecks: { [rowToUpdate.id]: isMuted } } }` would work better, but this is probably fine.

**Review by mook-as — APPROVED** [review]
LGTM, seems to work. (We never delete settings, but that's probably okay)
